File: leetcode/golang/LCR_181_fan_zhuan_dan_ci_shun_xu_lcof.go
func reverseMessage(message string) string {
    j := len(message) - 1
    var res []byte
    for j >= 0 && message[j] == ' ' {
        j--
    }
    for j >= 0 {
        end := j
        for j >= 0 && message[j] != ' ' {
            j--
        }
        for i := j + 1; i <= end; i++ {
            res = append(res, message[i])
        }
        res = append(res, ' ')
        for j >= 0 && message[j] == ' ' {
            j-- 
        }
    }
    if len(res) > 0 && res[len(res) - 1] == ' ' {
        res = res[: len(res) - 1]
    }
    return string(res)
}

File: leetcode/c++/LCR_181_fan_zhuan_dan_ci_shun_xu_lcof.cpp
// https://leetcode.cn/problems/fan-zhuan-dan-ci-shun-xu-lcof/description

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    string reverseMessage(string message) {
        int j = message.length() - 1, end;
        string res;
        while (j >= 0 && message[j] == ' ') j--;
        while (j >= 0) {
            end = j;
            while (j >= 0 && message[j] != ' ') j--;
            for (int i = j + 1; i <= end; i++)
                res.push_back(message[i]);
            res.push_back(' ');
            while (j >= 0 && message[j] == ' ') j--;
        }
        if (!res.empty() && res.back() == ' ') {
            res.pop_back();
        }
        return res;
    }
};

// O(n), O(n)